filebeat/tests/system/test_json.py

```diff
@@ -410,3 +410,63 @@ def test_integer_condition(self):
         output = self.read_output()
         assert len(output) == 1
         assert output[0]["status"] == 404
+
+    def test_keep_original(self):
+        """
+        JSON events contain raw version if json.keep_original_message is set
+        """
+
+        self.render_config_template(
+            path=os.path.abspath(self.working_dir) + "/log/*",
+            json=dict(
+                keys_under_root=True,
+                keep_original_message=True,
+            )
+        )
+
+        os.mkdir(self.working_dir + "/log/")
+        self.copy_files(["logs/json_int.log"],
+                        target_dir="log")
+
+        proc = self.start_beat()
+        self.wait_until(
+            lambda: self.output_has(lines=2),
+            max_timeout=10)
+        proc.check_kill_and_wait()
+
+        output = self.read_output()
+        assert len(output) == 2
+        assert all("log.original" in o for o in output)
+
+    def test_keep_original_multiline(self):
+        """
+        Multiline JSON events contain raw version if json.keep_original_message is set
+        under the configured json.original_message_target field
+        """
+
+        self.render_config_template(
+            path=os.path.abspath(self.working_dir) + "/log/*",
+            json=dict(
+                message_key="log_line",
+                keys_under_root=True,
+                keep_original_message=True,
+                original_message_target="my_custom_original_message",
+            ),
+            multiline=True,
+            pattern="^\[",
+            negate="true",
+            match="after",
+        )
+
+        os.mkdir(self.working_dir + "/log/")
+        self.copy_files(["logs/json_multiline.log"],
+                        target_dir="log")
+
+        proc = self.start_beat()
+        self.wait_until(lambda: self.output_has(lines=1))
+        proc.check_kill_and_wait()
+
+        output = self.read_output()
+        assert len(output) == 1
+        assert "my_custom_original_message" in output[0]
+        assert output[0]["my_custom_original_message"].count("\n") == 2
```

Comment:
What we have here is what I would expect in `event.original`. In `log.original I would expect that it is already the combined json.

Reply:
I am not sure I understand what you mean by your comment. This is the event what is produced by the Beat in this test:

```
{
  "@timestamp": "2018-11-22T15:26:02.702Z",
  "@metadata": {
    "beat": "filebeat",
    "type": "doc",
    "version": "7.0.0"
  },
  "log": {
    "offset": 0,
    "file": {
      "path": "/home/n/go/src/github.com/elastic/beats/filebeat/build/system-tests/run/test_json.Test.test_keep_original_multiline/log/json_multiline.log"
    },
    "flags": [
      "multiline"
    ]
  },
  "json": {},
  "input": {
    "type": "log"
  },
  "host": {
    "name": "sleipnir"
  },
  "agent": {
    "type": "filebeat",
    "hostname": "sleipnir",
    "version": "7.0.0"
  },
  "message": "[2018-05-05 18:00] line 1\n    line 2\n    line 3\n",
  "my_custom_original_message": "{\"log_line\": \"[2018-05-05 18:00] line 1\"}\n{\"log_line\": \"    line 2\"}\n{\"log_line\": \"    line 3\"}"
}
```
Three raw events are joined by a newline.

Do you want to see something else instead?

Reply:
I was thinking that it should already be a valid JSON object. What you have above I would expect for a multiline event but not in the case of docker. I think combining the different json objects into one is something that is very effectively done on the edge. It is a modification of the event but all the data is still "original", ready for processing.

Reply:
You mean something like this as the value of `log.original`:
```
{
    {\"log_line\": \"[2018-05-05 18:00] line 1\"},
    {\"log_line\": \"    line 2\"},
    {\"log_line\": \"    line 3\"}
}
```
?

Reply:
I was more thinking of `{\"log_line\": \"[2018-05-05 18:00] line 1\n    line 2\n    line 3\"},` In this case it's almost identical to the message but encoded json string is still there.

Reply:
I do see that this format has its advantages, e.g. no need for multiline parsing during reprocessing. But I am worried about the performance overhead of merging multiple events with numerous (nested) fields. I think I would prefer to have raw events concatenated under `log.original` and have it parsed later again IF needed. I would not do any parsing in advance if I do not know for sure that I am going to need it.

Reply:
My thinking on what should go into log.original so far: Everything that is done by the input / readers should be done and then put into log.original. This means in the ingest pipeline could in general rely and processing / reprocessing log.original. In the above case I think the multiline part is local knowledge of the input and not on processing side.

The statement about ingest pipelines relying on log.original has a bit bigger implications like we should potentially not do dissect or json decoding on the Beats side. But I would ignore the local processors for now and focus on what the input does.

What you currently have for log.original above for me would belong into event.original.